// @sap-ux/adp-tooling/dist/writer/writer-config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getConfig = getConfig;
const client_1 = require("../client");
const project_utils_1 = require("./project-utils");
const descriptor_content_1 = require("./descriptor-content");
/**
 * Generates the configuration object for the Adaptation Project.
 *
 * @param {ConfigOptions} options - The configuration options.
 * @param {AbapServiceProvider} options.provider - The ABAP service provider instance.
 * @param {ConfigAnswers} options.configAnswers - User-provided configuration details (system, application, etc.).
 * @param {FlexLayer} options.layer - The FlexLayer indicating the deployment layer.
 * @param {object} options.defaults - Default project parameters.
 * @param {string} options.defaults.namespace - The default namespace to be used.
 * @param {PackageJson} options.packageJson - The package.json information for generating custom configuration.
 * @param {ToolsLogger} options.logger - The logger for debugging and error logging.
 * @returns {Promise<AdpWriterConfig>} A promise that resolves to the generated ADP writer configuration.
 */
async function getConfig(options) {
    const { configAnswers, defaults, layer, logger, packageJson, provider } = options;
    const ato = await provider.getAtoInfo();
    const operationsType = ato.operationsType ?? 'P';
    const target = await (0, client_1.getAbapTarget)(configAnswers.system, logger);
    const customConfig = (0, project_utils_1.getCustomConfig)(operationsType, packageJson);
    return {
        app: {
            id: defaults.namespace,
            reference: configAnswers.application.id,
            layer,
            title: '',
            content: [(0, descriptor_content_1.getNewModelEnhanceWithChange)()]
        },
        customConfig,
        target,
        options: {
            fioriTools: true,
            enableTypeScript: false
        }
    };
}
